### ISSUE TITLE
[CI] testSecurityActionsByLicenseType: Failed to execute phase [query]

### ISSUE THREAD

**Issue by DaveCTurner** [elastic/elasticsearch#30301]
In https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.3+multijob-unix-compatibility/os=fedora/16/console, `org.elasticsearch.license.LicensingTests#testSecurityActionsByLicenseType` failed. The `REPRODUCE LINE` doesn't reproduce a failure:

```
./gradlew :x-pack:plugin:security:test -Dtests.seed=993121463E9AC0EF -Dtests.class=org.elasticsearch.license.LicensingTests -Dtests.method="testSecurityActionsByLicenseType" -Dtests.security.manager=true -Dtests.locale=sr -Dtests.timezone=Africa/Accra
```

This stack trace appears, but there's nothing obviously related to it in the rest of the logs:

```
Failed to execute phase [query], all shards failed
  at __randomizedtesting.SeedInfo.seed([993121463E9AC0EF:5C5ECE7262D69883]:0)
  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:288)
  at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:128)
  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:249)
  at org.elasticsearch.action.search.InitialSearchPhase.onShardFailure(InitialSearchPhase.java:101)
  at org.elasticsearch.action.search.InitialSearchPhase.access$100(InitialSearchPhase.java:48)
  at org.elasticsearch.action.search.InitialSearchPhase$2.lambda$onFailure$1(InitialSearchPhase.java:222)
  at org.elasticsearch.action.search.InitialSearchPhase.maybeFork(InitialSearchPhase.java:176)
  at org.elasticsearch.action.search.InitialSearchPhase.access$000(InitialSearchPhase.java:48)
  at org.elasticsearch.action.search.InitialSearchPhase$2.onFailure(InitialSearchPhase.java:222)
  at org.elasticsearch.action.search.SearchExecutionStatsCollector.onFailure(SearchExecutionStatsCollector.java:73)
  at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:51)
  at org.elasticsearch.action.search.SearchTransportService$ConnectionCountingHandler.handleException(SearchTransportService.java:527)
  at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1095)
  at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1188)
  at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1172)
  at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:66)
  at org.elasticsearch.action.search.SearchTransportService$6$1.onFailure(SearchTransportService.java:385)
  at org.elasticsearch.search.SearchService$2.onFailure(SearchService.java:341)
  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:335)
  at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:329)
  at org.elasticsearch.search.SearchService$3.doRun(SearchService.java:1019)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:724)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  at org.elasticsearch.common.util.concurrent.TimedRunnable.doRun(TimedRunnable.java:41)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
```

**Comment by DaveCTurner** [comment]
Here are the logs related to this test: [consoleText.txt.gz](https://github.com/elastic/elasticsearch/files/1963668/consoleText.txt.gz)

**Comment by jdconrad** [comment]
Another (https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/1859/console)

**Comment by jaymode** [comment]
Another failure here, https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+intake/1711/console

**Comment by jaymode** [comment]
I pushed a change to increase logging around search to see exactly what is causing all of the shards to fail during the search.

**Comment by danielmitterdorfer** [comment]
Another instance of this on 6.2 in https://internal-ci.elastic.co/job/elastic+x-pack-elasticsearch+6.2+matrix-java-periodic/ES_BUILD_JAVA=java9,ES_RUNTIME_JAVA=java8,nodes=linux/94/console

reproduction line:

```
./gradlew :x-pack-elasticsearch:plugin:security:test \
  -Dtests.seed=D37D8894A94058BA \
  -Dtests.class=org.elasticsearch.license.LicensingTests \
  -Dtests.method="testSecurityActionsByLicenseType" \
  -Dtests.security.manager=true \
  -Dtests.locale=fi \
  -Dtests.timezone=Africa/Porto-Novo
```

[build log](https://github.com/elastic/elasticsearch/files/2068321/build.log.zip)

**Comment by jaymode** [comment]
I haven't been able to find a reproduction with the increased logging, but I did spot how there could be a issue with the test. The test relies on updating the internal license state of a node but does not have any protection against a new change overwriting the changed license state.

Additionally, this test also triggers a NoNodeAvailableException when a node's client is trying to talk to itself. The underlying cause is an Authentication failure due to the license state changing in the middle of authentication. A token gets extracted but cannot be authenticated because the realm is no longer returned by the `Realms` method. While this would be rare in production scenarios, this is a real bug.

I've opened #38815 to address these items and will have that PR close this issue. If we see failures, we can re-open or create a new issue.
